### ISSUE TITLE
Code review model

### ISSUE THREAD

**Issue by zbraniecki** [unicode-org/icu4x#95]
I remember we talked about this on several occasions, but I don't think we ever clearly defined the approach to reviewing we want to take.

For a bit I've been experimenting with one, that I'd like to suggest for us:

> Every PR requires at least one review to be merged. 
> 
> Merging should be performed by the PR author. If the author wants to grant another team member rights to merge, they can state so in the PR comment.
> 
> Review can be of either architectural or syntactical value, and often is of both.
> The PR author can specify, when requesting review, what kind of review they are asking for, from each reviewer or even which area they'd like the reviewer to focus on (useful when distributing reviews).
>
> ## Reviewers role
>
> The reviewer is responsible for accepting only once they feel the current PR with their comments *not applied* is ready to be merged. In other words, the `accepted` can be set with pending review comments, if those comments don't affect whether the patch is ready to be merged (for example, they're stylistic suggestions).
> 
> The reviewer is also responsible for indicating the nature of their review comments - specifically, between three types: "blocking", "suggestion", "optional". `Blocking` is when the reviewer considers the change to be unmergable and requires a new revision. `Suggestion` is for when the reviewer considers the change to be suboptimal, but usable, and wants to defer the decision to the PR author, while stating their opinion. `Suggestion` is for when the reviewer considers multiple options to be mostly comparable or tradeoffs, and wants to defer to the PR author for the final decision after bringing up a new option.
> 
> In particular, reviewer should try to minimize their impact on the PR, and use the `blocking` conservatively. Their role is not to "improve the code" or "teach" the PR author, but rather help the PR author avoid making mistakes and lowering the quality of the code.
In result most review comments should be in form of suggestions and optionals, and the reviewer should request another revision for review only if they feel the need to see it before they can evaluate if the PR is mergable.
> 
> Lastly, the reviewers role is to evaluate the stakeholders group and ensure that the review coverage is complete - if they review only portion of the PR, or if they see the need for additional stakeholders to be involved, they should add additional reviewers, or CC them into the issue, depending on what kind of involvement they expect (inform vs verify).
> 
> ## Selection of reviewers
> 
> When the PR author creates a new PR, they should consider three sources of reviewers and CCed stakeholders:
> 
> * Owners and peers of the component they work with
> * People involved in the preceeding conversation
> * Recognized experts in the domain the PR operates in
> 
> The goal of the PR author is to find the subset of stakeholders that represent those three groups well. Depending on the scope and priority of the PR, the reviewers group size can be adjusted, with small PRs being sufficent for review by just one stakeholder, and larger PRs, or first-of-a-kind using a larger pool of reviewers.
> 
> ## PR author and reviewers workflow
> 
> When the PR author submits the PR for review, and once they selected the initial group, each member of that group can chose to either review or resign from reviewing.
If they resign, their only duty is to ensure that there is a sufficient coverage of the review, and nominate their replacement if not.
> 
> After a round of review, if there are blocking issues or a reviewer kept themselves in the review queue without accepting a review, the PR author must update the PR and re-request review from all pending reviewers.
> 
> PRs should generally not be merged with pending reviewers.
>
> If the PR author decides to make any substantial changes that go beyond of what the reviewers already approved, they can re-request an already accepted review after updating the PR.

This is a braindump based on my experience working with Mozilla Bugzilla, Github, ECMA402 and other projects.
I expect the text itself to require a rewrite, but first, I'd like to collect feedback on the content.
I tried to fit it into RACI (informed vs consulted) and plug `review required`, `input required` vs `sign-off required`, but Github kind of flattens this so it leaves the comments and the PR author to state what kind of input is requested and the only distinction we can draw is between CC (inform) and review request (put on some form of a blocking path).

Let's discuss this!

**Comment by zbraniecki** [comment]
@sffc @nciric @Manishearth @hsivonen @echeran @hagbard @filmil @kpozin @mihnita

**Comment by sffc** [comment]
Generally sounds okay.

I'm not sure about the distinction between blocking, suggested, and optional comments.  I think drawing that distinction downplays the importance of "nitpick" comments that, if ignored on a routine basis, could reduce the overall code quality.  I think the reviewer should leave comments; the author should either address them or leave a comment saying why they aren't going to address them; and the reviewer should either accept the author's conclusions (pressing the approve button), or leave further comments.  No need to distinguish between comment severity.

> Github kind of flattens this so it leaves the comments and the PR author to state what kind of input is requested and the only distinction we can draw is between CC (inform) and review request (put on some form of a blocking path)

We can use a bot and GitHub API status checks to add extra features if needed.

Can we also consider the best practices when using the "resolve" button on a PR review?  In ICU, we use the convention that once a comment is addressed, the PR author should leave a "Done" comment, and the reviewer should come back and either press the "resolve" button or leave a further comment if the issue was not fully addressed.

**Comment by zbraniecki** [comment]
>  I think drawing that distinction downplays the importance of "nitpick" comments that, if ignored on a routine basis, could reduce the overall code quality.

That's a good point. I wasn't sure how to locate the `consistency` comments, which, as you pointed out, accrue over time and lead to a lower quality of the project.
I'd like to find a balance in wording that says "don't reject the PR because someone used different loop than you like", but "let's make sure our code remains consistent".
Maybe a way out would be to state that the PR author is responsible for applying all consistency related comments.

> No need to distinguish between comment severity.

In my experience it is sometimes hard to distinguish that from reviewers wording if they're suggesting an improvement or not accepting my approach.

I'm ok softening the wording in this part to indicate more of a "Reviewer should try to be clear about how their feedback should be treated" or something like this.

> We can use a bot and GitHub API status checks to add extra features if needed.

Not sure how it could be used to indicate the difference between `I'm asking sffc for a generic architectural sign-off, mihai for the architectural review of the public API bits, and manish for the detail review of the memory management`.

> Can we also consider the best practices when using the "resolve" button on a PR review? In ICU, we use the convention that once a comment is addressed, the PR author should leave a "Done" comment, and the reviewer should come back and either press the "resolve" button or leave a further comment if the issue was not fully addressed.

I like that! At Mozilla we often use `thumbs up` to indicate that we applied the feedback, but it's the same as `Done`. In Phabricator we have `Done` checkbox.

**Comment by zbraniecki** [comment]
I'm not sure how it would work tho, if I accepted your PR with comments, you applied those comments and commented "Done". Will I know to come back and press "Resolve" on each, or will you have to poke me about it?
But we can iterate over it as we go :)

**Comment by sffc** [comment]
> Not sure how it could be used to indicate the difference between I'm asking sffc for a generic architectural sign-off, mihai for the architectural review of the public API bits, and manish for the detail review of the memory management.

We could add fields in the PR description, like,

```
ARCHITECTURE=sffc
API=mihnita
MEMORY=Manishearth
```

The bot can then have a status check that is green only if there is at least one approval from each category.  I set up something similar with ICU; you can put `ALLOW_MULTIPLE_COMMITS` in the PR description to disable the check that fails if you have multiple commits in your PR.

> I'm not sure how it would work tho, if I accepted your PR with comments, you applied those comments and commented "Done". Will I know to come back and press "Resolve" on each, or will you have to poke me about it?

Good point.  My model assumes that you only approve the PR after all your comments have been addressed.  That does add one extra back-and-forth.

> I'm ok softening the wording in this part to indicate more of a "Reviewer should try to be clear about how their feedback should be treated" or something like this.

Is the objective simply to make it more clear to the reviewer which comments were intended as required and which ones were intended as optional?

I personally write comments like "Optional: blah blah blah" if they are optional.  The absence of the prefix "Optional:" implies that I see the comment as required.  I've seen other people use the "Nit:" prefix, which is also fine.

We could establish a best practice along those lines, without making any hard requirements.

**Comment by zbraniecki** [comment]
> My model assumes that you only approve the PR after all your comments have been addressed. That does add one extra back-and-forth.

Right, I'm not attached to the idea, but I like the concept that often I can approve the PR with comments, without having to get back and re-review whether you applied my comments.

> Is the objective simply to make it more clear to the reviewer which comments were intended as required and which ones were intended as optional?

Yeah! And your approach seems totally solving the distinction I wanted to have. The one missing piece is the third option "suggestion".
The way I think about it is:
* Required - I think this is a bug, wrong choice, and I don't feel comfortable approving it without changes.
* Suggestion - I have experience and knowledge that makes me suggest what I consider a better approach, but yours will work too.
* Option - Your solution is one of multiple, and here's another. I'm not sure which one is better, so I'm just putting it under your consideration and you can decide which tradeoff is the right one to take.

**Comment by mihnita** [comment]
In my work reviews I only have 2 kind of comments:* default: should be either implemented, or a "why not" explanation should be providedI point out the reason for request (consistency, style guides, best practices, etc.)* nitpick: implement if you like it, ignore it otherwise. I don't expect an answer.If all my open comments are nitpicks I also approve, I don't wait for another loop.On May 15, 2020 14:04, Zibi Braniecki <notifications@github.com> wrote:

My model assumes that you only approve the PR after all your comments have been addressed. That does add one extra back-and-forth.

Right, I'm not attached to the idea, but I like the concept that often I can approve the PR with comments, without having to get back and re-review whether you applied my comments.

Is the objective simply to make it more clear to the reviewer which comments were intended as required and which ones were intended as optional?

Yeah! And your approach seems totally solving the distinction I wanted to have. The one missing piece is the third option "suggestion".
The way I think about it is:
Required - I think this is a bug, wrong choice, and I don't feel comfortable approving it without changes.Suggestion - I have experience and knowledge that makes me suggest what I consider a better approach, but yours will work too.Option - Your solution is one of multiple, and here's another. I'm not sure which one is better, so I'm just putting it under your consideration and you can decide which tradeoff is the right one to take.

**Comment by echeran** [comment]
I think the discussion already resolves most of my comments around the original "Reviewers Role" section pretty well.  

One thing I wonder is what the landscape of Rust tooling is like to automate the production of comments/warnings according to style guides and conventions.  For Java, regarding syntactic concerns, there are [style checkers](https://github.com/checkstyle/checkstyle) and [formatters](https://github.com/google/google-java-format), and Rust has `cargo fmt`.  Regarding slightly higher-level concerns about programming constructs, etc that are also mentioned in style guides, there are also such [programming construct checkers](https://github.com/google/error-prone). Is there an equivalent Rust to the ErrorProne checker for Java?

**Comment by zbraniecki** [comment]
`cargo clippy` ?

**Comment by echeran** [comment]
Ahh, okay.  I'll create a bug for that, based on discussion in #94.

**Comment by zbraniecki** [comment]
It seems that the current settings for master branch require re-review after a push, which makes the "approve with comment" model not viable.

We can change that to not stale approvals on new commit and I'd prefer that anticipating a churn to require reviewers to come back after they already r+'ed with nits. (see #100 for an example).

**Comment by sffc** [comment]
We can make that change to the repo settings; the risk of course is that it will allow potentially unreviewed code into the repo.  I don't believe anyone here will be malicious, but it could help prevent, e.g., accidentally pushing a print statement, a `git merge` that accidentally pulls in more changes than intended, etc.  In ICU, we have even stricter rules in place in order to prevent the latter case, since it's easy to make mistakes in git.

**Comment by zbraniecki** [comment]
> I don't believe anyone here will be malicious, but it could help prevent, e.g., accidentally pushing a print statement, a git merge that accidentally pulls in more changes than intended, etc

I doubt that requesting re-review is the best way to avoid that. `cargo fmt`/`cargo clippy` should handle that.
Humans are not the best tool to avoid those kind of issues and asking them to re-review after they gave r+ is more likely to increase the workload than to actually bring value.
